compiler/qsc_rca/src/applications.rs

```diff
@@ -0,0 +1,665 @@
+                panic!("only input parameters are expected");
+            };
+
+            // If a dynamic application is provided, set the compute kind associated to the parameter accordingly.
+            let mut compute_kind = ComputeKind::Classical;
+            if let Some((dynamic_param_index, dynamic_param_value_kind)) = dynamic_param {
+                if input_param_index == dynamic_param_index {
+                    compute_kind = ComputeKind::Quantum(QuantumProperties {
+                        runtime_features: RuntimeFeatureFlags::empty(),
+                        value_kind: dynamic_param_value_kind,
+                    });
+                }
+            }
+
+            locals_map.insert(
+                local_var_id,
+                LocalComputeKind {
+                    local,
+                    compute_kind,
+                },
+            );
+        }
+        Self {
+            locals_map,
+            active_dynamic_scopes: Vec::new(),
+            return_expressions: Vec::new(),
+            return_type: return_type.clone(),
+            blocks: FxHashMap::default(),
+            stmts: FxHashMap::default(),
+            exprs: FxHashMap::default(),
+        }
+    }
+
+    fn close(self) -> ApplicationInstanceComputeProperties {
+        // Determine the value kind of the application instance by going through each return expression aggregating
+        // their value kind (if any).
+        let mut value_kinds = Vec::<ValueKind>::new();
+        for expr_id in self.return_expressions {
+            let expr_compute_kind = self
+                .exprs
+                .get(&expr_id)
+                .expect("expression compute kind should exist");
+            if let ComputeKind::Quantum(quantum_properties) = expr_compute_kind {
+                value_kinds.push(quantum_properties.value_kind);
+            }
+        }
+
+        // The application instance has a value kind only if at least one of its return expressions was quantum.
+        let value_kind = if value_kinds.is_empty() {
+            None
+        } else {
+            let initial_value_kind = if let Ty::Array(_) = self.return_type {
+                ValueKind::Array(RuntimeKind::Static, RuntimeKind::Static)
+            } else {
+                ValueKind::Element(RuntimeKind::Static)
+            };
+            let value_kind = value_kinds.iter().fold(
+                initial_value_kind,
+                |aggregated_value_kind, current_value_kind| {
+                    aggregated_value_kind.aggregate(*current_value_kind)
+                },
+            );
+            Some(value_kind)
+        };
+
+        ApplicationInstanceComputeProperties {
+            blocks: self.blocks,
+            stmts: self.stmts,
+            exprs: self.exprs,
+            value_kind,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct LocalsComputeKindMap(FxHashMap<LocalVarId, LocalComputeKind>);
+
```

Comment:
Since this uses `LocalVarId` and those are low numbered and (mostly) continguous, this could potentially be an `IndexMap` instead of `FxHashMap` for some perf gains.

Reply:
Good catch! Using `IndexMap` instead of `FxHashMap` now.